File: leds.cpp
// solution to r/dailyprogrammer challenge #290 [Intermediate] Blinking LEDs
// https://www.reddit.com/r/dailyprogrammer/comments/5as91q/20161102_challenge_290_intermediate_blinking_leds/

#include <iostream>
#include <string>
#include <cassert>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <cstdint>
#include <sstream>
#include <iomanip>

/* grammar of our mini programming language:

      <line>: <whitespace> <instruction> |
              <label>                    |
              <empty>

      <instruction> : ld a,<num> |
                      ld b,<num> |
                      out (0),a  |
                      rlca       |
                      rrca       |
                      djnz <labelref>             
*/

typedef enum {
    LOAD_A = 0,
    LOAD_B = 1,
    OUT = 2,
    RLCA = 3,
    RRCA = 4,
    DJNZ = 5,
    END = 6
} Instruction_t; // TODO: better name

struct Instruction {
    Instruction(Instruction_t type, std::uint8_t data) : type(type), data(data) {
		assert((type == LOAD_A) || (type == LOAD_B) || (type == DJNZ)); 
		// TODO: would be nice to separate LOAD from DJNZ...
    }
    Instruction(Instruction_t type) : type(type), data(0u) {
		assert((type == OUT) || (type == RLCA) ||(type == RRCA) || (type == END));
    }
    const Instruction_t type;
    std::uint8_t get_load_value(void) {
		assert((type == LOAD_A) || (type == LOAD_B));
		return data;
    }
    std::uint8_t get_label(void) {
		assert(type == DJNZ);
		return data;
    }
	void print(void) const {
		switch (type) {
		case LOAD_A: std::cout << "ld a," << (int)data << std::endl;		break;
		case LOAD_B: std::cout << "ld b," << (int)data << std::endl;		break;
		case OUT: std::cout << "out (0),a" << std::endl;					break;
		case RLCA: std::cout << "rlca" << std::endl;						break;
		case RRCA: std::cout << "rrca" << std::endl;						break;
		case DJNZ: std::cout << "djnz " << (int)data << std::endl;			break;
		case END: std::cout << "END" << std::endl;							break;
		default: std::cout << "ILLEGAL INSTRUCTION :" << type << std::endl; break; // should be unreachable
		}
	}
private:   
    const std::uint8_t data; // value to load A or B, or label index
			                 // note: assuming B is 8 bits (not clear from problem), also assuming there will be a max of 255 labels
};

void run(std::string file_name);
std::string leds_to_string(uint8_t register_a);

std::string leds_to_string(uint8_t register_a) {
	std::string str;
	for (int i = 0; i < 8; i++) {
		uint8_t led = register_a & (0x80u >> i);
		if (led == 0) {
			str.push_back('.'); // 0 = Off
		}
		else {
			str.push_back('*'); // 1 = On
		}
	}
	return str;
}

// TODO: split up this function?
void run(std::string file_name) {   
	std::vector<Instruction> instructions;
	std::unordered_map<std::string, std::uint8_t> label_indexes; // TODO: spelling?

	// parse the file
	// TODO: try std::regex instead
	std::ifstream in_file(file_name);
	if (!in_file.is_open()) {
		std::cout << "Error: could not open file '" << file_name << "'" << std::endl;
	}

	std::string line;
	while (std::getline(in_file, line)) {
		// process one line at a time
		size_t index = line.find_first_not_of(" \n\r\t");
		if (index == std::string::npos) {
			// empty, skip
		}
		else {
			std::string str = line.substr(index, line.length());
			if (str == "rlca") {
				instructions.push_back(Instruction(RLCA));
			}
			else if (str == "rrca") {
				instructions.push_back(Instruction(RRCA));
			}
			else if (str == "out (0),a") {
				instructions.push_back(Instruction(OUT));
			}
			else {
				std::string s = str.substr(0, 5);
				if (s == "ld a,") {
					int a;
					std::istringstream(str.substr(5, str.size())) >> a;
					if ((a >= 0) && (a <= 255)) {
						instructions.push_back(Instruction(LOAD_A, a));
					}
				}
				else if (s == "ld b,") {
					int b;
					std::istringstream(str.substr(5, str.size())) >> b;
					if ((b >= 0) && (b <= 255)) {
						instructions.push_back(Instruction(LOAD_B, b));
					}
				}
				else if (s == "djnz ") {
					std::string label = str.substr(5, str.size()); // rest of the string
					int index = label_indexes[label]; // TODO: what if label doesnt exist? 
					// note: problem description guarantees jumps to label will always be backwards - so label should already exist
					instructions.push_back(Instruction(DJNZ, index));
				}
				else {
					// try to read string as a label
					// note: technically according to the problem description labels are not allowed to have whitespace in front of them but oh well
					int colon_index = str.find_first_of(':');
					if (colon_index != std::string::npos) {
						// TODO: not as strict as problem description for label, fix later
						std::string label = str.substr(0, colon_index);
						label_indexes.insert({ label, instructions.size() }); // TODO: what if label is after last instruction?
					}
					else {
						// no : found, cant be a label
						std::cout << "error parsing line, skipping:   " << line << std::endl;
					}
				}
			}
		}
		
    }
    instructions.push_back(Instruction(END));
   
	// print the file
	//for (const Instruction& instr : instructions) {
	//instr.print();
	//}

    // run the file
    int instruction_index = 0;
    uint8_t a;
    uint8_t b;
	bool done = false;
	while (!done) {
		Instruction next = instructions[instruction_index];
		instruction_index++;

		switch (next.type) {
		case LOAD_A:
			a = next.get_load_value();
			break;

		case LOAD_B:
			b = next.get_load_value();
			break;

		case OUT:
			std::cout << leds_to_string(a) << std::endl;
			break;

		case RLCA: {
			uint8_t temp = a >> 7;
			a = (a << 1) | temp;
			break;
		}
		case RRCA: {
			uint8_t temp = a << 7;
			a = (a >> 1) | temp;
			break;
		}
		case DJNZ: 
			// dont decrement if already 0
			if (b > 0u) {
				b--;
			}
			// if still not 0 after decrement, jump
			if (b > 0u) {
				instruction_index = next.get_label();
			}
			break;

		case END: 
			done = true;
			break;

		default: 
			std::cout << "ILLEGAL INSTRUCTION :" << next.type << std::endl;
			done = true; // this should be unreachable - end immediately if something gones this wrong
			break;
		}
	}
}

int main(void) {
   //run("input1.txt");
   //run("input2.txt");
   //run("input3.txt");
   run("input4.txt");
   
   return 0;
}

File: leds.rs
// solution to r/dailyprogrammer challenge #290 [Intermediate] Blinking LEDs
// https://www.reddit.com/r/dailyprogrammer/comments/5as91q/20161102_challenge_290_intermediate_blinking_leds/
// port of the c++ version

use std::fs::File;
use std::io::prelude::*;
use std::collections::HashMap;

/* grammar of our mini programming language:

      <line>: <whitespace> <instruction> |
              <label>                    |
              <empty>

      <instruction> : ld a,<num> |
                      ld b,<num> |
                      out (0),a  |
                      rlca       |
                      rrca       |
                      djnz <labelref>             
*/

#[derive(Debug, PartialEq)]
enum Line {
    Instruction( Instruction ),
    Label( String ),
    Empty,
    ParseError
}

#[derive(Debug, PartialEq)]
enum Instruction {
    LoadA( u8 ),
    LoadB( u8 ),
    Out,
    Rlca,
    Rrca,
    Djnz{ index: usize },
}

impl Instruction {
    #[allow(dead_code)]
    fn to_string(&self) -> String {
        match self {
            &Instruction::LoadA( a ) => format!("ld a,{}", a),
            &Instruction::LoadB( b ) => format!("ld b,{}", b),
            &Instruction::Out => String::from("out (0),a"),
            &Instruction::Rlca => String::from("rlca"),
            &Instruction::Rrca => String::from("rrca"),
            &Instruction::Djnz{ index } => format!("djnz {}", index),
       }
    }
}

fn leds_to_string(register_a: u8) -> String {
    let mut result = String::new();
    for c in (0..8).map(|x| 0x80 >> x) {
        match c & register_a {
            0 => result.push('.'),
            _ => result.push('*'),
        }
    }
    result
}

fn parse(line: &str, labels: &HashMap<String, usize>) -> Line {
    // note: allowing the following even though they are technically not part of the problem description:
    //  line -> <whitespace> <instruction> <whitespace>
    //  line -> <whitespace> <label> <whitespace>

    let trimmed: &str = line.trim();
    match trimmed {
        "rlca" => Line::Instruction(Instruction::Rlca),
        "rrca" => Line::Instruction(Instruction::Rrca),
        "out (0),a" => Line::Instruction(Instruction::Out),
        "" => Line::Empty,
        _ => {
            if trimmed.len() < 5 {
                try_label(trimmed, labels)
            }
            else {
                match &trimmed[0..5] {
                    "ld a," => try_register_load(trimmed, true),
                    "ld b," => try_register_load(trimmed, false),
                    "djnz " => try_jump(trimmed, labels),
                    _ => try_label(trimmed, labels),
                }
            }
        },
    }
}

fn try_jump(trimmed: &str, labels: &HashMap<String, usize>) -> Line {    
    let new_label: &str = &trimmed[5..trimmed.len()];
    match labels.get(new_label) {
        Some(line) => Line::Instruction(Instruction::Djnz{ index: *line }),
        None => Line::ParseError,
    }
}

fn try_label(trimmed: &str, labels: &HashMap<String, usize>) -> Line {
    let mut temp = trimmed.to_owned();   
    match temp.pop() {
        // last char in trimmed line should be the semicolon
        Some(':') => {
            // label must be a valid format, and cannot match an existing label
            if is_identifier(&temp) && !labels.contains_key(&temp) {
                Line::Label(temp)
            }
            else {
                Line::ParseError
            }
        },
        _ => Line::ParseError,
    }
}

fn is_identifier(chars: &str) -> bool {
    // this function determines which characters I allow in the name of a label
    let mut ret: bool = true;
    for c in chars.chars() {
        if !c.is_alphanumeric() && c != '_' {
            ret = false;
            break;
        }
    }
    ret
}

fn try_register_load(trimmed: &str, register_a: bool) -> Line {
    // note: register_a == false means instruction is a LoadB
    let load_value = &trimmed[5..trimmed.len()].parse::<u8>();
    match (register_a, load_value) {
        (true, &Ok(a)) => Line::Instruction(Instruction::LoadA(a)),
        (false, &Ok(b)) => Line::Instruction(Instruction::LoadB(b)),
        (_, &Err(_)) => Line::ParseError,
    }
}

fn run_file(file_name: &str) {
    let mut file = File::open(file_name).expect(&(format!("Failed to open {}", file_name)));
    let mut contents = String::new();
    file.read_to_string(&mut contents).expect(&(format!("Failed to read from {}", file_name)));
    
    let mut instructions: Vec<Instruction> = vec![];
    let mut labels: HashMap<String, usize> = HashMap::new();

    // parse the input file into instructions
    for line in contents.lines() {
        match parse(line, &labels) {
            Line::Label(name) => { labels.insert(String::from(name), instructions.len()); },
            Line::Instruction(instruction) => instructions.push(instruction),
            Line::Empty => (),
            Line::ParseError => println!("Error parsing line:\n  '{}'", line),
        }
    }
    
    //for inst in instructions {
    //    println!("{}", inst.to_string());
    //}

    // execute the instructions
    let mut register_a: u8 = 0;
    let mut register_b: u8 = 0;
    let mut instruction_index: usize = 0;
    while instruction_index < instructions.len() {
        match instructions[instruction_index] {
            Instruction::LoadA(a) => register_a = a,
            Instruction::LoadB(b) => register_b = b,
            Instruction::Out => { println!("{}", leds_to_string(register_a)); },
            Instruction::Rlca => register_a = register_a.rotate_left(1),
            Instruction::Rrca => register_a = register_a.rotate_right(1),
            Instruction::Djnz{index: i} => {
                if register_b > 0 {
                    register_b -= 1;
                }                
                if register_b > 0 {
                    instruction_index = i;
                    continue; // don't want to add 1 in this case, so skip that part
                }
            },
        }
        instruction_index += 1;
    }
}

fn main() {
    //run_file("input1.txt");
    //run_file("input2.txt");
    //run_file("input3.txt");
    run_file("input4.txt");
}

#[test]
#[allow(dead_code)]
fn test_try_register_load() {
    assert_eq!(try_register_load("ld a,4", true), Line::Instruction(Instruction::LoadA(4)));
    assert_eq!(try_register_load("ld a,4", true), Line::Instruction(Instruction::LoadA(4)));
    assert_eq!(try_register_load("ld b,4", false), Line::Instruction(Instruction::LoadB(4)));
    assert_eq!(try_register_load("ld a,400", true), Line::ParseError);
    assert_eq!(try_register_load("ld a,a123", true), Line::ParseError);
    assert_eq!(try_register_load("ld a,", true), Line::ParseError);
}

#[test]
#[allow(dead_code)]
fn test_try_label() {
    let labels: HashMap<String, usize> = HashMap::new();
    assert_eq!(try_label("lbl:", &labels), Line::Label("lbl".to_owned())); // String::from() ?
    assert_eq!(try_label("longer_label:", &labels), Line::Label("longer_label".to_owned()));
    assert_eq!(try_label("bad:label:", &labels), Line::ParseError);
    assert_eq!(try_label("toomanycolons::", &labels), Line::ParseError);
    assert_eq!(try_label("foo", &labels), Line::ParseError);
    assert_eq!(try_label(":bar", &labels), Line::ParseError);
}

#[test]
#[allow(dead_code)]
fn test_try_jump() {
    let mut labels: HashMap<String, usize> = HashMap::new();
    labels.insert("label".to_string(), 0);
    assert_eq!(try_jump("djnz label", &labels), Line::Instruction(Instruction::Djnz{ index: 0 }));
}

#[test]
#[allow(dead_code)]
fn test_parse() {
    let labels: HashMap<String, usize> = HashMap::new();

    // make sure correct boolean is supplied to try_register_load()
    assert_eq!(parse("  ld a,4", &labels), Line::Instruction(Instruction::LoadA(4)));
    assert_eq!(parse("ld b,4   ", &labels), Line::Instruction(Instruction::LoadB(4)));
    
    // test instructions that don't have a dedicated function
    assert_eq!(parse("    ", &labels), Line::Empty);
    assert_eq!(parse(" rlca  ", &labels), Line::Instruction(Instruction::Rlca));
    assert_eq!(parse("  rrca ", &labels), Line::Instruction(Instruction::Rrca));
    assert_eq!(parse(" out (0),a ", &labels), Line::Instruction(Instruction::Out));
    
    // note: whitespace handling is covered in above tests
}

#[test]
#[allow(dead_code)]
fn test_leds_to_string() {
    assert_eq!(leds_to_string(0x80), "*.......");
    assert_eq!(leds_to_string(0x55), ".*.*.*.*");
    assert_eq!(leds_to_string(0xAA), "*.*.*.*.");
    assert_eq!(leds_to_string(0x0F), "....****");
}

